src/routes/Attendance.ts

```diff
@@ -0,0 +1,299 @@
+
+import { FieldSet, Records } from "airtable";
+
+import StudentDao from "@daos/Airtable/StudentDao";
+import CanvasCoursesDao from "@daos/Airtable/CanvasCoursesDao";
+import ModulesDao from "@daos/Canvas/ModulesDao";
+import { paramMissingError } from "@shared/constants";
+import SubmissionDao from "@daos/Canvas/SubmissionDao";
+import Module from "@entities/Module";
+import Submission from "@entities/Submission";
+import ModuleItem from "@entities/ModuleItem";
+
+const studentDao = new StudentDao();
+const canvasCoursesDao = new CanvasCoursesDao();
+const modulesDao = new ModulesDao();
+const submissionDao = new SubmissionDao();
+
+const { BAD_REQUEST, CREATED, OK } = StatusCodes;
+
+interface IMilestone {
+  name: string;
+  itemId: number;
+  points: number;
+  completed: boolean;
+}
+
+/**
+ * Get a set of learners' student records from Airtable by their email.
+ *
+ * @param learners
+ * @returns
+ */
+async function getStudentRecords(
+  learners: Record<string, unknown>[]
+): Promise<Records<FieldSet>> {
+  const emails: string[] = learners.map((learner) => learner.Email as string);
+  return await studentDao.getByEmails(emails);
+}
+
+/**
+ * Merge each learner's Lambda ID and Labs Role from their student record.
+ * (Mutates learners array.)
+ *
+ * @param learners
+ * @returns
+ */
+function mergeStudentRecords(
+  learners: Record<string, unknown>[],
+  studentRecords: Records<FieldSet>
+): Record<string, unknown>[] {
+  for (const learner of learners) {
+    const airtableRecord = studentRecords.find(
+      (elm) => elm.fields["Email"] == learner["Email"]
+    );
+    if (!airtableRecord) {
+      learner["Lambda ID"] = null;
+      learner["Labs Role"] = null;
+    } else {
+      learner["Lambda ID"] = airtableRecord.fields["Lambda ID"];
+      learner["Labs Role"] = airtableRecord.fields["Labs Role"];
+    }
+  }
+
+  return learners;
+}
+
+/**
+ * Look up a learner's Objectives course ID in Airtable by their Labs role.
+ *
+ * @param labsRole
+ * @returns
+ */
+async function getObjectivesCourseId(labsRole: string): Promise<number | null> {
+  return await canvasCoursesDao.getObjectiveCourseByRole(labsRole);
+}
+
+/**
+ * Get whether a sprint milestone in Canvas is marked complete.
+ *
+ * @param courseId
+ * @param assignmentId
+ * @param learnerId
+ * @returns
+ */
+async function getMilestoneCompleted(
+  courseId: number,
+  assignmentId: number,
+  learnerId: string
+): Promise<boolean | null> {
+  const userSubmission = await submissionDao.getByAssignmentAndUser(
+    courseId,
+    assignmentId,
+    learnerId
+  );
+
+  return userSubmission ? userSubmission.grade === "complete" : null;
+}
+
+/**
+ * Get the assignment submission for the given event.
+ *
+ * @param learnerId
+ * @param eventType
+ * @param courseId
+ * @param moduleId
+ * @returns
+ */
+async function getMilestone(
+  learnerId: string,
+  eventType: string,
+  courseId: number,
+  module: Module
+): Promise<IMilestone | null> {
```

Comment:
Why are you returning promises on an async function?

Reply:
This could go on the canvas DAO along with many of the other non-router funcitons.

Reply:
Wasn't sure about this one—it's accessing multiple DAOs. Should we create a `services` dir?

Reply:
[Added a service layer](https://github.com/Lambda-School-Labs/labs-be/pull/14/commits/cb9e0b0acfeec828a74d8196bbb03d573832b01c)